src/main/scala/org/apache/spark/shuffle/ucx/ShuffleTransport.scala

```diff
@@ -0,0 +1,126 @@
+ */
+trait Cookie
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: MemoryBlock
+}
+
+
+object OperationStatus extends Enumeration {
+  val SUCCESS, FAILURE = Value
+}
+
+/**
+ * Operation statistic, like completionTime, transport used, protocol used, etc.
+ */
+trait OperationStats {
+  def getElapsedTimeNs: Long
+}
+
+class TransportError(errorMsg: String) extends Exception(errorMsg)
+
+trait OperationResult {
+  def recvSize: Long
+  def getStatus: OperationStatus.Value
+  def getError: TransportError
+  def getStats: OperationStats
+}
+
+/**
+ * Async operation callbacks
+ */
+trait OperationCallback {
+  def onComplete(result: OperationResult)
+}
+
+/**
+ * Transport flow example:
+ * val transport = new UcxShuffleTransport()
+ * transport.init()
+ *
+ * Mapper/writer:
+ * transport.register(metadataBlockId, metadataBlock) // we don't care for metadata cookies
+ *
+ * Reducer:
+ * transport.fetchBlockByBlockId(blockId, resultBounceBuffer)
+ *
+ * transport.unregister(blockId)
+ * transport.close()
+ */
+trait ShuffleTransport {
+
+  /**
+   * Initialize transport resources. This function should get called after ensuring that SparkConf
+   * has the correct configurations since it will use the spark configuration to configure itself.
+   * @return worker address of current process, to use in [[ addExecutor()]]
+   */
+  def init(): ByteBuffer
+
+  /**
+   * Close all transport resources
+   */
+  def close()
+
+  /**
+   * Add executor's worker address. For standalone testing purpose and for implementations that makes
+   * connection establishment outside of UcxShuffleManager.
+   */
+  def addExecutor(executorId: String, workerAddress: ByteBuffer)
+
+  /**
+   * Registers blocks using blockId on SERVER side.
+   */
+  def register(blockId: BlockId, block: Block): Cookie
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  def mutate(blockId: BlockId, block: Block, callback: OperationCallback)
+
+  /**
+   * Indicate that this blockId is not needed any more by an application
+   */
+  def unregister(blockId: BlockId)
+
+  /**
+   * Fetch remote blocks by blockIds.
+   */
+  def fetchBlockByBlockId(executorId: String, blockIds: BlockId,
+                          resultBuffer: MemoryBlock, cb: OperationCallback)
+
+  /**
+   * Fetch remote blocks by cookies.
+   */
+  def fetchBlocksByCookies(executorId: String, cookies: Seq[Cookie],
+                           resultBuffer: MemoryBlock, cb: OperationCallback)
+
+  /**
+   * Progress outstanding operations. This routine is blocking. It's important to call this routine
+   * within same thread that submitted requests.
+   */
+  def progress()
```

Comment:
So the flow is likely:

1. [executor task thread] On `next()`, call `fetchBlocksByCookies` and then call `progress`
2. I should then wait (?) for the callback, so I need potentially a queue or a condition variable, and this task can't proceed obviously, has no data. The callback thread is going to be (?) We need to document in the api what the expectation is for the thread that is invoking the callback (can I block it?, is it a critical thread?)
3. When callback is called, the condition (memory is available) is met and `next()` can return. 
4. If the callback isn't called, I need to call `progress()` also in [executor task thread] some more?

Can we discuss more exactly how `progress()` works? This should be documented in the function definition as well.

Reply:
https://github.com/petro-rudenko/sparkucx/blob/public-transport-api-impl/src/main/scala/org/apache/spark/shuffle/ucx/perf/UcxShuffleTransportPerfTool.scala#L154

Reply:
Note to self: the thread calling the callback is the task thread that called `progress`

Reply:
1. If elements are in completion queue, `progress` calls callback.
2. If no elements in completion queue, `progress` returns 0 (@petro-rudenko we need to add a return code). So we'll need to busy wait?
3. We should use the wait/event driven api (the current impl in rapids shuffle uses _wakeup_ from UCX), s.t. we don't busy wait, since burning a core is not acceptable, especially with many potential calls to `progress` that are all waiting on congested TCP.
4. So can we hide `progress` if it's not too hard, and have a config that can choose the wakeup (epoll) option.